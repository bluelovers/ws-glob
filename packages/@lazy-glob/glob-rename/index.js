"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.globRenameAsync = exports.globRenameSync = exports.globRenameResult = void 0;
const tslib_1 = require("tslib");
const globRenameResult_1 = require("./lib/globRenameResult");
Object.defineProperty(exports, "globRenameResult", { enumerable: true, get: function () { return globRenameResult_1.globRenameResult; } });
const defaultRenameHandler_1 = require("./lib/defaultRenameHandler");
tslib_1.__exportStar(require("./lib/types"), exports);
function globRenameSync(searchGlobPattern, resultGlobList, options) {
    const { handler = defaultRenameHandler_1.defaultRenameHandler } = options;
    const resultList = (0, globRenameResult_1.globRenameResult)(searchGlobPattern, resultGlobList, options);
    for (const result of resultList) {
        handler(result, options);
    }
}
exports.globRenameSync = globRenameSync;
async function globRenameAsync(searchGlobPattern, resultGlobList, options) {
    const { handler = defaultRenameHandler_1.defaultRenameHandler } = options;
    const resultList = (0, globRenameResult_1.globRenameResult)(searchGlobPattern, resultGlobList, options);
    for await (const result of resultList) {
        await handler(result, options);
    }
}
exports.globRenameAsync = globRenameAsync;
exports.default = globRenameSync;
//# sourceMappingURL=index.js.map