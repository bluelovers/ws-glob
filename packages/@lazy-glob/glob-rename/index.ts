import { globRenameResult } from './lib/globRenameResult';
import { IResultGlobListRow, IOptionsGlobRenameResult, IOptionsGlobRename } from './lib/types';
import { defaultRenameHandler } from './lib/defaultRenameHandler';

export * from './lib/types';

export { globRenameResult }

export function globRenameSync(searchGlobPattern: string | string[],
	resultGlobList: IResultGlobListRow | IResultGlobListRow[],
	options: IOptionsGlobRename,
)
{
	const { handler  = defaultRenameHandler } = options

	const resultList = globRenameResult(searchGlobPattern, resultGlobList, options);

	for (const result of resultList)
	{
		handler(result, options)
	}
}

export async function globRenameAsync(searchGlobPattern: string | string[],
	resultGlobList: IResultGlobListRow | IResultGlobListRow[],
	options: IOptionsGlobRename,
)
{
	const { handler = defaultRenameHandler } = options

	const resultList = globRenameResult(searchGlobPattern, resultGlobList, options);

	for await (const result of resultList)
	{
		await handler(result, options)
	}
}

export default globRenameSync
