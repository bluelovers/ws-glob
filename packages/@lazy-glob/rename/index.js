"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeName = void 0;
const computeName_1 = require("./lib/util/computeName");
Object.defineProperty(exports, "computeName", { enumerable: true, get: function () { return computeName_1.computeName; } });
exports.default = computeName_1.computeName;
//# sourceMappingURL=index.js.map