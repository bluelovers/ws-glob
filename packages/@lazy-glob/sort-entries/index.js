"use strict";
/**
 * Created by user on 2020/6/9.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.sort = exports.entries_reduce = exports.entries_sort = void 0;
const string_natural_compare_1 = require("@bluelovers/string-natural-compare");
const util_1 = require("@lazy-glob/util");
function entries_sort(entries, fn = string_natural_compare_1.naturalCompare, cache = {}) {
    entries = entries.reduce(function (a, b) {
        let v = b[1];
        if (v === null || typeof v == 'string') {
            a.push(b);
        }
        else {
            let d = Object.entries(v);
            a.push([b[0], entries_sort(d, fn), v[util_1.SymGlobTree]]);
        }
        return a;
    }, []);
    entries.sort(function (a, b) {
        let r = fn(a[0], b[0], cache);
        return r;
    });
    return entries;
}
exports.entries_sort = entries_sort;
function entries_reduce(entries) {
    return entries
        .reduce(function (a, [k, v, bool]) {
        if (v === null || typeof v == 'string') {
            a[k] = v;
        }
        else {
            a[k] = entries_reduce(v);
        }
        if (bool) {
            a[k][util_1.SymGlobTree] = bool;
        }
        return a;
    }, {});
}
exports.entries_reduce = entries_reduce;
function sort(a, fn = string_natural_compare_1.naturalCompare) {
    let r = Object.entries(a);
    let a1 = entries_sort(r, fn);
    let a2 = entries_reduce(a1);
    return a2;
}
exports.sort = sort;
exports.default = sort;
//# sourceMappingURL=index.js.map