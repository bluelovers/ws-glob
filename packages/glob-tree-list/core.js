"use strict";
/**
 * Created by user on 2018/3/29/029.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const upath2_1 = require("upath2");
exports.path = upath2_1.default;
function globToTree(data) {
    return data.reduce(function (a, b) {
        b = upath2_1.default.normalize(b);
        let dirname = upath2_1.default.dirname(b);
        let basename = upath2_1.default.basename(b);
        let isdir = b.slice(-1) == upath2_1.default.sep;
        if (isdir) {
            basename += upath2_1.default.sep;
        }
        //console.log([dirname, basename]);
        if (dirname == '.') {
            let f = a;
            f[basename] = isdir ? null : basename;
        }
        else {
            let c = dirname
                .split(upath2_1.default.sep);
            if (c[0] == '.') {
                c.shift();
            }
            let f = a;
            c.forEach(function (e) {
                e += upath2_1.default.sep;
                f[e] = f[e] || {};
                f = f[e];
            });
            f[basename] = isdir ? (f[basename] || {}) : basename;
        }
        return a;
    }, {});
}
exports.globToTree = globToTree;
function treeToGlob(a, d = []) {
    return Object.entries(a).reduce(function (a, b) {
        //console.log(b);
        if (b[1] === null || typeof b[1] == 'string') {
            let k = (b[1] === null ? b[0] : b[1]);
            if (d.length) {
                // @ts-ignore
                a.push(upath2_1.default.join(...d, k));
            }
            else {
                a.push(k);
            }
        }
        else {
            // @ts-ignore
            a = a.concat(treeToGlob(b[1], d.concat(b[0])));
        }
        return a;
    }, []);
}
exports.treeToGlob = treeToGlob;
globToTree.globToTree = globToTree;
globToTree.treeToGlob = treeToGlob;
globToTree.default = globToTree;
exports.default = exports;
//# sourceMappingURL=core.js.map