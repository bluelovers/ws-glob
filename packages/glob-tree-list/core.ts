/**
 * Created by user on 2018/3/29/029.
 */

import path from 'upath2';
export { path }

export function globToTree(data: string[]): ITree
{
	return data.reduce(function (a, b)
	{
		b = path.normalize(b);
		let dirname = path.dirname(b);
		let basename = path.basename(b);

		let isdir = b.slice(-1) == path.sep;

		if (isdir)
		{
			basename += path.sep;
		}

		//console.log([dirname, basename]);

		if (dirname == '.')
		{
			let f = a;

			f[basename] = isdir ? null : basename;
		}
		else
		{
			let c = dirname
				.split(path.sep)
			;

			if (c[0] == '.')
			{
				c.shift();
			}

			let f = a;

			c.forEach(function (e)
			{
				e += path.sep;

				f[e] = f[e] || {};

				f = f[e];
			});

			f[basename] = isdir ? (f[basename] || {}) : basename;
		}

		return a;
	}, {});
}

export function treeToGlob(a: ITree, d: string[] = []): string[]
{
	return Object.entries(a).reduce(function (a, b)
	{
		//console.log(b);

		if (b[1] === null || typeof b[1] == 'string')
		{
			let k = (b[1] === null ? b[0] : b[1]) as string;

			if (d.length)
			{
				// @ts-ignore
				a.push(path.join(...d, k));
			}
			else
			{
				a.push(k);
			}
		}
		else
		{
			// @ts-ignore
			a = a.concat(treeToGlob(b[1], d.concat(b[0])));
		}

		return a;
	}, [] as string[]);
}

export module globToTree
{
	export type ITreeEntries = string | ITreeRow | null;

	export type ITreeRow = {
		[key: string]: ITreeEntries;
	}

	export type ITree = {
		[key: string]: string | ITreeRow | null;
	}
}

export import ITreeEntries = globToTree.ITreeEntries;
export import ITreeRow = globToTree.ITreeRow;
export import ITree = globToTree.ITree;

globToTree.globToTree = globToTree;
globToTree.treeToGlob = treeToGlob;
globToTree.default = globToTree;

export default exports as typeof import('./core');
