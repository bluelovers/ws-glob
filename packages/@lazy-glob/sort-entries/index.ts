/**
 * Created by user on 2020/6/9.
 */

import { naturalCompare } from '@bluelovers/string-natural-compare';
import { ITree } from '@lazy-glob/util/lib/types';
import { ISortEntriesSort, ISortEntriesSortRow } from '@lazy-glob/util/lib/types/entries';
import { SymGlobTree } from '@lazy-glob/util';

export function entries_sort(entries, fn: (a: string, b: string, cache) => number = naturalCompare, cache: any = {}): ISortEntriesSort
{
	entries = entries.reduce(function (a: ISortEntriesSort, b: ISortEntriesSortRow)
	{
		let v = b[1];

		if (v === null || typeof v == 'string')
		{
			a.push(b);
		}
		else
		{
			let d = Object.entries(v);
			a.push([b[0], entries_sort(d, fn), v[SymGlobTree]]);
		}

		return a;
	}, [] as ISortEntriesSort);

	(entries as ISortEntriesSort).sort(function (a, b)
	{
		let r = fn(a[0], b[0], cache);
		return r;
	});

	return entries;
}

export function entries_reduce(entries)
{
	return (entries as ISortEntriesSort)
		.reduce(function (a, [k, v, bool]: ISortEntriesSortRow)
		{
			if (v === null || typeof v == 'string')
			{
				a[k] = v as string;
			}
			else
			{
				a[k] = entries_reduce(v);
			}

			if (bool)
			{
				a[k][SymGlobTree] = bool;
			}

			return a;
		}, {} as ITree)
		;
}

export function sort(a: ITree, fn: (a: string, b: string, cache) => number = naturalCompare)
{
	let r = Object.entries(a);

	let a1 = entries_sort(r, fn);
	let a2 = entries_reduce(a1);

	return a2
}

export default sort
