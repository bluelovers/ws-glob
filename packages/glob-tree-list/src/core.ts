/**
 * Created by user on 2018/3/29/029.
 */

//export * from './lib/types'
import { SymGlobTree } from '@lazy-glob/util';

export * from '@lazy-glob/util/lib/types';
import path from 'upath2';

import { ITree } from '@lazy-glob/util/lib/types';
export { path }

export function globToTree(data: string[]): ITree
{
	return data.reduce(function (a, b)
	{
		b = path.normalize(b);
		let dirname = path.dirname(b);
		let basename = path.basename(b);

		let isdir = b.slice(-1) == path.sep;

		if (isdir)
		{
			basename += path.sep;
		}

		//console.log([dirname, basename]);

		if (dirname == '.')
		{
			let f = a;

			f[basename] = isdir ? null : basename;
		}
		else
		{
			let c = dirname
				.split(path.sep)
			;

			if (c[0] == '.')
			{
				c.shift();
			}

			let f = a;

			c.forEach(function (e)
			{
				e += path.sep;

        // @ts-ignore
				f[e] = f[e] || {};

				// @ts-ignore
				f = f[e];
			});

			f[basename] = isdir ? (f[basename] || {}) : basename;

			if (isdir)
			{
        // @ts-ignore
				f[basename][SymGlobTree] = true;
				//console.dir({ b, basename, f })
			}
		}

		return a;
	}, {} as ITree);
}

export function treeToGlob(a: ITree, d: string[] = []): string[]
{
	return Object.entries(a).reduce(function (a, b)
	{
		//console.log(b);

		if (b[1] === null || typeof b[1] == 'string')
		{
			let k = (b[1] === null ? b[0] : b[1]) as string;

			if (d.length)
			{
				// @ts-ignore
				a.push(path.join(...d, k));
			}
			else
			{
				a.push(k);
			}
		}
		else
		{
			let ls = treeToGlob(b[1], d.concat(b[0]));

			if (b[1][SymGlobTree])
			{
				let k = b[0];

				if (d.length)
				{
					// @ts-ignore
					a.push(path.join(...d, k));
				}
				else
				{
					a.push(k);
				}
			}

			a = a.concat(ls);
		}

		return a;
	}, [] as string[]);
}

globToTree.globToTree = globToTree;
globToTree.treeToGlob = treeToGlob;
globToTree.default = globToTree;

export default exports as typeof import('./core');

