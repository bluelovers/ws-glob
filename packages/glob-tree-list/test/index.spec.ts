
import { globToTree, treeToGlob } from '../core';
import { sort } from '@lazy-glob/sort-entries';
import naturalCompare from '@bluelovers/string-natural-compare/core';

let data = `00020_1章.txt
00020_2章/
00020_3章/
00020_3章/3章 9話/3章 9話file.txt
00020_3章/3章 8話.txt
00020_3章/3章 10話.txt
00020_3章/3章 11話.txt
00020_3章/3章 12話.txt
00020_3章/3章 13話.txt
00020_3章/3章 15話.txt
00020_3章/3章 14話.txt
00020_3章/3章 15.5話 特別閑話.txt
00020_3章/3章 16話.txt
00020_3章/3章 17話.txt`.split("\n");

test(`globToTree / treeToGlob`, () =>
{
	let t = globToTree(data);

	let actual = sort(t, function (a, b, cache)
	{
		return naturalCompare(_c(a, cache), _c(b, cache));
	});

	expect(t).toMatchSnapshot();
	expect(actual).toMatchSnapshot();
	expect(treeToGlob(actual)).toMatchSnapshot();

});

function _c(k, cache): string
{
	if (k in cache)
	{
		return cache[k];
	}

	cache[k] = k
		.replace(/\D+$/, '')
	;

//	console.log(k, cache[k]);

	return cache[k];
}
