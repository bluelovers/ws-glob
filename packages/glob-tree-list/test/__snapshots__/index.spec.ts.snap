// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`globToTree / treeToGlob 1`] = `
{
  "00020_1章.txt": "00020_1章.txt",
  "00020_2章/": null,
  "00020_3章/": {
    "3章 10話.txt": "3章 10話.txt",
    "3章 11話.txt": "3章 11話.txt",
    "3章 12話.txt": "3章 12話.txt",
    "3章 13話.txt": "3章 13話.txt",
    "3章 14話.txt": "3章 14話.txt",
    "3章 15.5話 特別閑話.txt": "3章 15.5話 特別閑話.txt",
    "3章 15話.txt": "3章 15話.txt",
    "3章 16話.txt": "3章 16話.txt",
    "3章 17話.txt": "3章 17話.txt",
    "3章 8話.txt": "3章 8話.txt",
    "3章 9話/": {
      "3章 9話file.txt": "3章 9話file.txt",
    },
  },
}
`;

exports[`globToTree / treeToGlob 2`] = `
{
  "00020_1章.txt": "00020_1章.txt",
  "00020_2章/": null,
  "00020_3章/": {
    "3章 10話.txt": "3章 10話.txt",
    "3章 11話.txt": "3章 11話.txt",
    "3章 12話.txt": "3章 12話.txt",
    "3章 13話.txt": "3章 13話.txt",
    "3章 14話.txt": "3章 14話.txt",
    "3章 15.5話 特別閑話.txt": "3章 15.5話 特別閑話.txt",
    "3章 15話.txt": "3章 15話.txt",
    "3章 16話.txt": "3章 16話.txt",
    "3章 17話.txt": "3章 17話.txt",
    "3章 8話.txt": "3章 8話.txt",
    "3章 9話/": {
      "3章 9話file.txt": "3章 9話file.txt",
    },
  },
}
`;

exports[`globToTree / treeToGlob 3`] = `
[
  "00020_1章.txt",
  "00020_2章/",
  "00020_3章/3章 8話.txt",
  "00020_3章/3章 9話/3章 9話file.txt",
  "00020_3章/3章 10話.txt",
  "00020_3章/3章 11話.txt",
  "00020_3章/3章 12話.txt",
  "00020_3章/3章 13話.txt",
  "00020_3章/3章 14話.txt",
  "00020_3章/3章 15話.txt",
  "00020_3章/3章 15.5話 特別閑話.txt",
  "00020_3章/3章 16話.txt",
  "00020_3章/3章 17話.txt",
]
`;
